colossus/src/main/scala/colossus/controller/Pipe.scala

```diff
@@ -0,0 +1,411 @@
+  }
+
+  /**
+   * Cancels execution of the trigger.  The pusher should only do this when they're about to terminate the stream.
+   *
+   * TODO: might be a better way to handle this, but it would probably imply we'd need to know who terminated the stream
+   */
+  def cancel() {
+    callback = None
+  }
+
+}
+
+sealed trait PushResult {
+  def pushed: Boolean
+}
+object PushResult {
+  sealed trait Pushed extends PushResult {
+    def pushed = true
+  }
+  sealed trait NotPushed extends PushResult {
+    def pushed = false
+  }
+
+  //the item was successfully pushed and is ready for more data
+  case object Ok extends Pushed
+
+  //the item was pushed, but is approaching capacity, proactive backpressure measures should be taken if possible
+  case object Filling extends Pushed
+
+  //the item was successfully pushed but the pipe is not yet ready for more data, trigger is called when it's ready
+  case class Filled(onReady: Signal) extends Pushed
+
+  //the item was not pushed because the pipe is already full, the same trigger
+  //returned when the Filled result was returned is included
+  case class Full(onReady: Signal) extends NotPushed
+
+  //The pipe has been manually closed (without error) and is not accepting any more items
+  case object Closed extends NotPushed
+
+  //The pipe has been terminated or some other error has occurred
+  case class Error(reason: Throwable) extends NotPushed
+}
+
+/**
+ * Wraps 2 sinks and will automatically begin reading from the second only when
+ * the first is empty.  The `None` from the first sink is never exposed.  The
+ * first error reported from either sink is propagated.
+ */
+class DualSource[T](a: Source[T], b: Source[T]) extends Source[T] {
+  private var a_empty = false
+  def pull(cb: Try[Option[T]] => Unit) {
+    if (a_empty) {
+      b.pull(cb)
+    } else {
+      a.pull{
+        case Success(None) => {
+          a_empty = true
+          pull(cb)
+        }
+        case other => cb(other)
+      }
+    }
+  }
+
+  def terminate(reason: Throwable) {
+    a.terminate(reason)
+    b.terminate(reason)
+  }
+
+  override def terminated: Boolean = if (a_empty) b.terminated else a.terminated
+
+  def isClosed = a.isClosed && b.isClosed
+}
+
+
+
+sealed trait PipeException extends Throwable
+class PipeTerminatedException(reason: Throwable) extends Exception("Pipe Terminated", reason) with PipeException
+class PipeStateException(message: String) extends Exception(message) with PipeException
+
+/**
+ * This is a special exception that Input/Output controllers look for when
+ * error handling pipes.  In most cases they will log the error that terminated
+ * the pipe, but for this one exception, the failure will be silent.  This is
+ * basically for situations where a certain amount of data is expected but for
+ * some reason the receiver decides to cancel for some business-logic reason.
+ */
+class PipeCancelledException extends Exception("Pipe Cancelled") with PipeException
```

Comment:
Yeah it's kind of arbitrary now.  I made it a regular class since you don't want an object as an exception and a case class with no parameters seems weird.  I do think this does bring forth an issue that exception for pipes probably need to be a bit more structured and unified.